packages/utils/odsp-doclib-utils/src/resolveWithDomainMove.ts

```diff
@@ -0,0 +1,66 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { IRequest } from "@fluidframework/core-interfaces";
+import { IUrlResolver } from "@fluidframework/driver-definitions";
+import { IOdspResolvedUrl, OdspErrorType } from "@fluidframework/odsp-driver-definitions";
+
+/**
+ * Checks if the error is domain move error.
+ * @param error - error whose type is to be determined.
+ * @returns - True is the error is domain move error.
+ */
+ export function isDomainMoveError(error: any) {
+    if (typeof error === "object" && error !== null
+        && error.errorType === OdspErrorType.locationRedirection) {
+        return true;
+    }
+    return false;
+}
+
+/**
+ * Handles odsp domain change handling while fulfilling the loader request.
+ * @param api - Callback in which user can wrap the loader.resolve or loader.request call.
+ * @param request - request to be resolved.
+ * @param urlResolver - resolver used to resolve the url.
+ * @returns - Response from the api call.
+ */
+export async function resolveWithDomainChangeHandling<T>(
+    api: (request: IRequest) => Promise<T>,
+    request: IRequest,
+    urlResolver: IUrlResolver,
+): Promise<T> {
+    let req: IRequest = request;
+    let success = false;
+    let response: T | undefined;
+    let odspResolvedUrl: IOdspResolvedUrl | undefined;
+    do {
+        try {
+            response = await api(req);
+            success = true;
+        } catch(error: any) {
+            if (isDomainMoveError(error)) {
+                // Need to generate only once.
+                if (!odspResolvedUrl) {
+                    odspResolvedUrl = await urlResolver.resolve(req) as IOdspResolvedUrl;
```

Comment:
What if
- Host uses MultiUrlResolver with both ODSP & FRS drivers / resolvers
- Both FRS & ODSP support redirects?
The issue here is that we assume that ODSP resolver was used, and I think it's a bit risky assumption.
I liked (in your second case) where we delegated this work of resolving / remapping back to URI resolver. I'd rather do that, but would make sure it's capture in interface signature, i.e. it becomes part of contract.

Alternatively, we can require drivers to provide this extra information on the actual domain Move error itself. Then we do not need to complicate uri resolver or have guesses - we can simply pick up new data and start over. I like it much better than leaking this info to various layers.

Reply:
This (suggestion) would also allow this code not to be ODSP specific and move out to some shared layer.

Reply:
I do not think it matters. This code should be generic enough and support any driver / storage that has or wants to add that capability.
Even if only ODSP supports it, I'd prefer this logic not to be in ODSP driver (or utils) and be in some shared place, and not have dirrect references to ODSP. I.e. design it in a way where adding same capability to another driver is trivial.

Reply:
I agree about putting it in a separate package. In that case, we can go with first option.

Reply:
Take a look at my earlier proposal - is it possible to return all required info on error object itself, such that URI resolver should not be minimally involved? I.e. can error object contain new resolved URL? That way we need to call getAbsolutIUrl, but that's fine - this function already exists, and driver does not know what format of unresolved URL app expects (i.e. there can be many resolvers for given driver). So driver knows how to patch resolved URL and it feels like it's driver responsibility to do so, and responsibility of this code to convert it back to unresolved URL using url resolver that host specified.